src/Moderation/DomainPatternBlocklist.php

```diff
@@ -0,0 +1,226 @@
+	}
+
+	public function get(): array
+	{
+		return $this->config->get('system', 'blocklist', []);
+	}
+
+	public function set(array $blocklist): bool
+	{
+		$result = $this->config->set('system', 'blocklist', $blocklist);
+		if ($result) {
+			$this->notifyAll();
+		}
+
+		return $result;
+	}
+
+	/**
+	 * @param string      $pattern
+	 * @param string|null $reason
+	 * @return int 0 if the block list couldn't be saved, 1 if the pattern was added, 2 if it was updated in place
+	 */
+	public function addPattern(string $pattern, string $reason = null): int
+	{
+		$update = false;
+
+		$blocklist = [];
+		foreach ($this->get() as $blocked) {
+			if ($blocked['domain'] === $pattern) {
+				$blocklist[] = [
+					'domain' => $pattern,
+					'reason' => $reason ?? self::DEFAULT_REASON,
+				];
+
+				$update = true;
+			} else {
+				$blocklist[] = $blocked;
+			}
+		}
+
+		if (!$update) {
+			$blocklist[] = [
+				'domain' => $pattern,
+				'reason' => $reason ?? self::DEFAULT_REASON,
+			];
+		}
+
+		$this->set($blocklist);
+
+		return $this->set($blocklist) ? ($update ? 2 : 1) : 0;
+	}
+
+	/**
+	 * @param string $pattern
+	 * @return int 0 if the block list couldn't be saved, 1 if the pattern wasn't found, 2 if it was removed
+	 */
+	public function removePattern(string $pattern): int
+	{
+		$found = false;
+
+		$blocklist = [];
+		foreach ($this->get() as $blocked) {
+			if ($blocked['domain'] === $pattern) {
+				$found = true;
+			} else {
+				$blocklist[] = $blocked;
+			}
+		}
+
+		return $found ? ($this->set($blocklist) ? 2 : 0) : 1;
+	}
+
+	public function exportToFile($filename)
+	{
+		$fp = fopen($filename, 'w');
+		if (!$fp) {
+			throw new Exception(sprintf('The file "%s" could not be created.', $filename));
+		}
+
+		foreach ($this->get() as $domain) {
+			fputcsv($fp, $domain);
+		}
+	}
+
+	public function append(array $blocklist)
+	{
+		$localBlocklist = $this->get();
+		$localPatterns  = array_column($localBlocklist, 'domain');
+
+		$importedPatterns = array_column($blocklist, 'domain');
+
+		$patternsToAppend = array_diff($importedPatterns, $localPatterns);
+
+		if (count($patternsToAppend)) {
+			foreach (array_keys($patternsToAppend) as $key) {
+				$localBlocklist[] = $blocklist[$key];
+			}
+
+			$this->set($localBlocklist);
+		}
+
+		return count($patternsToAppend);
+	}
+
+	public static function extractFromCSVFile($filename): array
+	{
+		$fp = fopen($filename, 'r');
+		if ($fp === false) {
+			throw new Exception(sprintf('The file "%s" could not be opened for importing', $filename));
+		}
+
+		$blocklist = [];
+		while (($data = fgetcsv($fp, 1000, ',')) !== false) {
+			$domain = $data[0];
+			if (count($data) == 0) {
+				$reason = self::DEFAULT_REASON;
+			} else {
+				$reason = $data[1];
+			}
+
+			$data = [
+				'domain' => $domain,
+				'reason' => $reason
+			];
+			if (!in_array($data, $blocklist)) {
+				$blocklist[] = $data;
+			}
+		}
+
+		return $blocklist;
+	}
+
+	public function notifyAll(): int
```

Comment:
Can you please describe what this method does?

Reply:
Not any better than its name suggests.

Reply:
Yes indeed.